apl-core/src/main/java/com/apollocurrency/aplwallet/apl/core/shard/ShardImporter.java

```diff
@@ -0,0 +1,141 @@
+import com.apollocurrency.aplwallet.apl.core.app.Blockchain;
+import com.apollocurrency.aplwallet.apl.core.app.BlockchainProcessor;
+import com.apollocurrency.aplwallet.apl.core.app.Genesis;
+import com.apollocurrency.aplwallet.apl.core.chainid.BlockchainConfig;
+import com.apollocurrency.aplwallet.apl.core.db.DatabaseManager;
+import com.apollocurrency.aplwallet.apl.core.db.DerivedTablesRegistry;
+import com.apollocurrency.aplwallet.apl.core.db.dao.ShardDao;
+import com.apollocurrency.aplwallet.apl.core.db.dao.model.Shard;
+import com.apollocurrency.aplwallet.apl.core.peer.DownloadableFilesManager;
+import com.apollocurrency.aplwallet.apl.core.shard.helper.CsvImporter;
+import com.apollocurrency.aplwallet.apl.util.Zip;
+import lombok.extern.slf4j.Slf4j;
+
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import javax.enterprise.inject.spi.CDI;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+@Singleton
+@Slf4j
+public class ShardImporter {
+    private ShardDao shardDao;
+
+    private DatabaseManager databaseManager;
+    private Blockchain blockchain;
+    private DerivedTablesRegistry derivedTablesRegistry;
+    private BlockchainConfig blockchainConfig;
+    private CsvImporter csvImporter;
+    private Zip zipComponent;
+    private DownloadableFilesManager downloadableFilesManager;
+    private AplAppStatus aplAppStatus;
+
+    @Inject
+    public ShardImporter(ShardDao shardDao, BlockchainConfig blockchainConfig, DatabaseManager databaseManager, Blockchain blockchain, DerivedTablesRegistry derivedTablesRegistry, CsvImporter csvImporter, Zip zipComponent, DownloadableFilesManager downloadableFilesManager, AplAppStatus aplAppStatus) {
+        this.shardDao = shardDao;
+        this.databaseManager = databaseManager;
+        this.blockchain = blockchain;
+        this.derivedTablesRegistry = derivedTablesRegistry;
+        this.csvImporter = csvImporter;
+        this.zipComponent = zipComponent;
+        this.downloadableFilesManager = downloadableFilesManager;
+        this.aplAppStatus = aplAppStatus;
+        this.blockchainConfig = blockchainConfig;
+
+    }
+
+    public void importShardByFileId(String fileId) {
+        importShard(fileId);
+        // set to start work block download thread (starting from shard's snapshot block here)
+        log.debug("Before updating BlockchainProcessor from Shard data and RESUME block downloading...");
+        BlockchainProcessor blockchainProcessor = CDI.current().select(BlockchainProcessor.class).get();
+        blockchainProcessor.updateInitialSnapshotBlock();
+        blockchainProcessor.resumeBlockchainDownloading(); // IMPORTANT CALL !!!
+    }
+
+    public void importLastShard(int height) {
+        if (height == 0) {
+            Genesis.apply(false);
+        } else {
+            Shard completedShard = shardDao.getLastCompletedShard();
```

Comment:
NPE can be here if there is not any shard exist

Reply:
This code should be called only when shard should  be imported and this shard already exists.